### ISSUE TITLE
Transpiler error during execution of circuits with dynamical decoupling on real backends

### ISSUE THREAD

**Issue by riddhisw** [Qiskit/qiskit-terra#7317]
### Environment

'qiskit-terra': '0.18.3', 'qiskit-aer': '0.9.1', 'qiskit-ignis': '0.6.0', 'qiskit-ibmq-provider': '0.18.1', 'qiskit-aqua': '0.9.5', 'qiskit': '0.32.1', 'qiskit-nature': None, 'qiskit-finance': None, 'qiskit-optimization': None, 'qiskit-machine-learning': None} 

### What is happening?

A Transpiler Error occurs when executing circuits with dynamical coupling on real backends:

`TranspilerError: "This circuit None may involve a delay instruction violating the pulse controller alignment. To adjust instructions to right timing, you should call one of scheduling passes first. This is usually done by calling transpiler with scheduling_method='alap'."`

Whether or not these errors occur to be dependent on the type of circuit. For instance, the following circuit executes without errors:

<img width="791" alt="image" src="https://media.github.ibm.com/user/319025/files/a12bee00-4ee9-11ec-97c0-57f06cb23599">

But the same test circuit FAILS  after adding one extra hadamard on qubit 0:
<img width="791" alt="image" src="https://media.github.ibm.com/user/319025/files/c3257080-4ee9-11ec-8016-ff216b3bffed">.

Commenting out the hadmard on qubit 0 will allow the circuit to execute without errors. Changing to `backend = FakeMumbai() ` also executes without errors. Changing `optimization level =1 `to other levels makes no difference. 
Code fails on multiple real devices (not just IBM mumbai).

### How can we reproduce the issue?

The code below 
```
from qiskit.transpiler import PassManager, InstructionDurations
from qiskit.transpiler.passes import ALAPSchedule, DynamicalDecoupling
from qiskit.circuit.library import XGate
from qiskit import QuantumCircuit, transpile, execute, IBMQ 

IBMQ.load_account()
provider = IBMQ.get_provider(hub='ibm-q-internal' , group='performance', project='demos')
realbackend = provider.get_backend('ibmq_mumbai')

circ = QuantumCircuit(4)
circ.cx(0, 1)
circ.cx(2, 3)
circ.h(0) # comment out to run without errors
circ.cx(1, 2)
circ.measure_all()

circ_t = transpile(circ, backend=realbackend,scheduling_method="alap")
dd_sequence = [XGate()] * 2
spacing = []
durations = InstructionDurations.from_backend(realbackend)
pm = PassManager([ALAPSchedule(durations),
                DynamicalDecoupling(durations, dd_sequence, qubits=None, spacing=None)])
circuits = pm.run(circ_t)

job = execute(circuits, realbackend, shots=10)
```

fails with error:

`TranspilerError: "This circuit None may involve a delay instruction violating the pulse controller alignment. To adjust instructions to right timing, you should call one of scheduling passes first. This is usually done by calling transpiler with scheduling_method='alap'."`

### What should happen?

Code should execute without Transpile errors irrespective of the type of circuit.

**Comment by mtreinish** [comment]
The issue here is in your code, specifically the use of `execute()` (which combines `transpile(qc, backend)` and `backend.run(qc)` into a single function call). What is happening is that you're scheduling your circuit with your custom pass manager first but not adjusting it to the timing constraints on the backend. Then when you call `execute()` again that's the equivalent of running `transpile(circuits, backend)` again without specifying a scheduling method. That `transpile()` call is raising an error telling you you've got a circuit with delays that doesn't comply with the scheduling constraints on the specified backend target and it won't run correctly without scheduling. The suggestion there is that the easiest way to fix this is to use the scheduling method kwarg on `transpile()` to run the full set of scheduling passes which will adjust the scheduled circuit based on the timing constraints. You have 2 ways you can fix this, either update your code to do:

```
circuits = pm.run(circuit_t)
job = realbackend.run(transpile(circuits, realbackend, scheduling_method="alap"), shots=10)
```

or add the timing constraint passes to your custom pass manager. You can see the relevant passes here:

https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/preset_passmanagers/level0.py#L213-L244

**Comment by riddhisw** [comment]
Thanks so much that worked!

**Comment by mtreinish** [comment]
Ok, then I'll close this issue. But please feel free to reopen it if there is more to discuss on this.

**Comment by riddhisw** [comment]
With the following code:

```

  realbackend = provider.get_backend(name)
  config = realbackend.configuration()
  timing_constraints = config.to_dict()['timing_constraints']
  print("Timing constraints for "+name)
  print(timing_constraints)
  print()

  circ = QuantumCircuit(4)
  circ.cx(0, 1)
  circ.cx(2, 3)
  circ.h(0) # comment out for no errors
  circ.cx(1, 2)
  circ.measure_all()

  circ_t = transpile(circ, backend=realbackend,scheduling_method="alap")
  dd_sequence = [XGate(),XGate()]
  spacing = []
  durations = InstructionDurations.from_backend(realbackend)
  pm = PassManager([ALAPSchedule(durations),
                  DynamicalDecoupling(durations, dd_sequence, qubits=None, spacing=None)])
  circuits = pm.run(circ_t)

  # circuits_dd = transpile(circuits, realbackend, scheduling_method="alap")
  # job = realbackend.run(circuits_dd, shots=10)

  try:
      job = realbackend.run(transpile(circuits, realbackend, scheduling_method="alap"), shots=10)
      print(job.result())
  except Exception as ecp:
      print(ecp)
```

I get 9999 internal error on mumbai.

```
ibmqfactory.load_account:WARNING:2021-12-07 20:13:54,619: Credentials are already in use. The existing account in the session will be replaced.
{'qiskit-terra': '0.18.3', 'qiskit-aer': '0.9.1', 'qiskit-ignis': '0.6.0', 'qiskit-ibmq-provider': '0.18.1', 'qiskit-aqua': '0.9.5', 'qiskit': '0.32.1', 'qiskit-nature': None, 'qiskit-finance': None, 'qiskit-optimization': None, 'qiskit-machine-learning': None}
Timing constraints for ibmq_mumbai
{'acquire_alignment': 16, 'granularity': 16, 'min_length': 64, 'pulse_alignment': 16}

'Unable to retrieve result for job 61b031091e11d5067a6bdcc4. Job has failed: Internal Error. Error code: 9999.'
```
I can change mumbai for other backends and I still get an error.

**Comment by riddhisw** [comment]
The following code

 ```
  realbackend = provider.get_backend(name)
  config = realbackend.configuration()
  timing_constraints = config.to_dict()['timing_constraints']
  print("Timing constraints for "+name)
  print(timing_constraints)
  print()


  circ = QuantumCircuit(4)
  circ.cx(0, 1)
  circ.cx(2, 3)
  circ.h(0) 
  circ.cx(1, 2)
  circ.measure_all()

  circ_t = transpile(circ, backend=realbackend,scheduling_method="alap")
  dd_sequence = [XGate(),XGate()]
  spacing = []
  durations = InstructionDurations.from_backend(realbackend)
  pm = PassManager([ALAPSchedule(durations),
                  DynamicalDecoupling(durations, dd_sequence, qubits=None, spacing=None)])
  circuits = pm.run(circ_t)

  try:
      job = realbackend.run(transpile(circuits, realbackend, scheduling_method="alap"), shots=10)
      print(job.result())
  except Exception as ecp:
      print(ecp)
```

produces the errors on IBM mumbai:
```

ibmqfactory.load_account:WARNING:2021-12-07 20:13:54,619: Credentials are already in use. The existing account in the session will be replaced.
{'qiskit-terra': '0.18.3', 'qiskit-aer': '0.9.1', 'qiskit-ignis': '0.6.0', 'qiskit-ibmq-provider': '0.18.1', 'qiskit-aqua': '0.9.5', 'qiskit': '0.32.1', 'qiskit-nature': None, 'qiskit-finance': None, 'qiskit-optimization': None, 'qiskit-machine-learning': None}
Timing constraints for ibmq_mumbai
{'acquire_alignment': 16, 'granularity': 16, 'min_length': 64, 'pulse_alignment': 16}

'Unable to retrieve result for job 61b031091e11d5067a6bdcc4. Job has failed: Internal Error. Error code: 9999.'
```

Other backends also yield the same error.

**Comment by nonhermitian** [comment]
The error coming back is 

`Exception: Delays must be a multiple of 16 samples`

due to the delay of `712dt` on q0

![image](https://user-images.githubusercontent.com/1249193/145189136-c3177dc4-1ca8-4cb5-bc7a-acd0c4e5e640.png)

**Comment by nkanazawa1989** [comment]
Note that you cannot change timing constraints on your side. This is constraints from the control electronics. If your backend says `pulse_alignment: 16` currently you cannot use DD pass. So far in all IBM systems this constraint has been 1 meaning you can place your instruction at arbitrary starting time. On the other hand, if this is not 1, the electronics will ask you to start every instruction at multiple of that value. Currently we only have such stretching pass for `acquire_alignment `
https://github.com/Qiskit/qiskit-terra/blob/d9ab98f2b3663b42ba354ed1cd2bc18b3f2cd19d/qiskit/transpiler/passes/scheduling/instruction_alignment.py#L29

Even if you add scheduling pass, no stretching to satisfy the constraint will happen without such alignment pass. However, current implementation of the stretching pass is really inefficient, since quantum circuit is too lazy for the timing context (they don't have t0 and we always need to scan all instructions in DAG to find t0). This may cause non-negligible performance regression in transpilation.

**Comment by nkanazawa1989** [comment]
The best (or temporarily) solution would be ask DD pass to check the constraints and insert delays that conform to the pulse alignment constraints. @ajavadia

**Comment by ajavadia** [comment]
Yeah I think the DD pass should be enhanced to take into account pulse constraints, if a backend reports them. Right now it is agnostic to that, so it can produce delays which violate the control electronics' constraints.

**Comment by pat-phattharaporn** [comment]
> The issue here is in your code, specifically the use of `execute()` (which combines `transpile(qc, backend)` and `backend.run(qc)` into a single function call). What is happening is that you're scheduling your circuit with your custom pass manager first but not adjusting it to the timing constraints on the backend. Then when you call `execute()` again that's the equivalent of running `transpile(circuits, backend)` again without specifying a scheduling method. That `transpile()` call is raising an error telling you you've got a circuit with delays that doesn't comply with the scheduling constraints on the specified backend target and it won't run correctly without scheduling. The suggestion there is that the easiest way to fix this is to use the scheduling method kwarg on `transpile()` to run the full set of scheduling passes which will adjust the scheduled circuit based on the timing constraints. You have 2 ways you can fix this, either update your code to do:
> 
> ```
> circuits = pm.run(circuit_t)
> job = realbackend.run(transpile(circuits, realbackend, scheduling_method="alap"), shots=10)
> ```
> 
> or add the timing constraint passes to your custom pass manager. You can see the relevant passes here:
> 
> https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/preset_passmanagers/level0.py#L213-L244

I tried this on my VQE circuit but it still raised that error. Any suggestion please? (here `ansatz_ = pm_dd.run(my_compiled_ansatz))`
<img width="758" alt="Screen Shot 2022-01-02 at 3 45 48 PM" src="https://user-images.githubusercontent.com/71392951/147892484-7d3eea26-902d-4ec1-bc00-01fed54e8d17.png">
This is the error:
<img width="999" alt="Screen Shot 2022-01-02 at 3 46 02 PM" src="https://user-images.githubusercontent.com/71392951/147892482-2a00adb4-37d4-4b10-9a8b-125d5fdb26ef.png">

**Comment by nonhermitian** [comment]
@nkanazawa1989 any progress to report here?  The above issue is a pain and hopefully we can squash it soon.

**Comment by nkanazawa1989** [comment]
Yes, this actually involves the refactoring of the whole framework in addition to the DD pass. I have updated the behavior of scheduling pass in #7655, and current stage is #7709 which upgrades scheduler pass and introducing `BasePadding` class which will be new base class of dynamical decoupling pass. Then, DD pass will be upgraded to consider the pulse alignment. Hopefully I can get reviews soon.

**Comment by ajavadia** [comment]
this should be fixed with #7760

**Comment by sirgeorgesawcon** [comment]
Hi, Is this issue closed, I am still facing the same error
![SmartSelect_20230824_010922_Chrome](https://github.com/Qiskit/qiskit/assets/63145298/50f317c2-b16e-44be-9c34-f6239b1026e0)



**steps to replicate the error.**

![SmartSelect_20230824_011227_Chrome](https://github.com/Qiskit/qiskit/assets/63145298/47c67bb5-e19d-489e-9549-a310b985e818)


This circuit is a part of Iterative Amplitude Estimation procedure.


![SmartSelect_20230824_011331_Chrome](https://github.com/Qiskit/qiskit/assets/63145298/064dc184-88e2-4671-87e0-d9246b26c2e2)



**qc is circ_ dd** here, basically at the end I made `qc = circ_dd`



From the above discussion, I get the key point that **add the timing constraint passes to your custom pass manager**, but how to do that? Since I can't fix it with simply by changing `execute` to `backend.run,`

Do I have to use anything else than ALAPScheduler?



 
![Web capture_24-8-2023_132246_qiskit org](https://github.com/Qiskit/qiskit/assets/63145298/bc7f47b5-5dba-4349-b694-20d40df6d94a)


Since Qiskit offers all these passes?
